database/raft/raft.go

```diff
@@ -0,0 +1,923 @@
+	b, err := snapshot.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	crc := crc32.Checksum(b, crcTable)
+	snap := snappb.Snapshot{Crc: crc, Data: b}
+	return snap.Marshal()
+}
+
+func decodeSnapshot(data []byte, snapshot *raftpb.Snapshot) error {
+	var snapPB snappb.Snapshot
+	err := snapPB.Unmarshal(data)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	if crc32.Checksum(snapPB.Data, crcTable) != snapPB.Crc {
+		return errors.Wrap(errors.New("bad snapshot crc"))
+	}
+	err = snapshot.Unmarshal(snapPB.Data)
+	return errors.Wrap(err)
+}
+
+func (sv *Service) applyEntry(ent raftpb.Entry, writers map[string]chan bool) {
+	log.Printkv(context.Background(), "index", ent.Index, "ent", ent)
+
+	switch ent.Type {
+	case raftpb.EntryConfChange:
+		var cc raftpb.ConfChange
+		err := cc.Unmarshal(ent.Data)
+		if err != nil {
+			panic(err)
+		}
+		sv.confState = *sv.raftNode.ApplyConfChange(cc)
+		switch cc.Type {
+		case raftpb.ConfChangeAddNode, raftpb.ConfChangeUpdateNode:
+			log.Printkv(context.Background(),
+				"what", cc.Type,
+				"id", cc.NodeID,
+				"addr", string(cc.Context),
+			)
+			sv.stateMu.Lock()
+			defer sv.stateMu.Unlock()
+			defer sv.stateCond.Broadcast()
+			sv.state.SetPeerAddr(cc.NodeID, string(cc.Context))
+		case raftpb.ConfChangeRemoveNode:
+			if cc.NodeID == sv.id {
+				log.Printkv(context.Background(), "nodeID", cc.NodeID, "msg", "removed from cluster")
+				panic(errors.New("removed from cluster"))
+			}
+			sv.stateMu.Lock()
+			defer sv.stateMu.Unlock()
+			defer sv.stateCond.Broadcast()
+			sv.state.RemovePeerAddr(cc.NodeID)
+		}
+	case raftpb.EntryNormal:
+		log.Printkv(context.Background(), "EntryNormal", ent)
+		//raft will send empty request defaulted to EntryNormal on leader election
+		//we need to handle that here
+		if ent.Data == nil {
+			break
+		}
+		sv.stateMu.Lock()
+		defer sv.stateCond.Broadcast()
+		defer sv.stateMu.Unlock()
+		var p proposal
+		err := json.Unmarshal(ent.Data, &p)
+		if err != nil {
+			log.Printf(context.Background(), "ent.data: %q", ent.Data)
+			panic(err)
+		}
+		satisfied, err := sv.state.Apply(p.Instruction, ent.Index)
+		if err != nil {
+			panic(err)
+		}
+		// send 'satisfied' over channel to caller
+		if c := writers[string(p.Wctx)]; c != nil {
+			c <- satisfied
+			delete(writers, string(p.Wctx))
+		}
+	default:
+		panic(fmt.Errorf("unknown entry type: %v", ent))
+	}
+}
+
+func (sv *Service) send(msgs []raftpb.Message) {
+	for _, msg := range msgs {
+		data, err := msg.Marshal()
+		if err != nil {
+			panic(err)
+		}
+		sv.stateMu.Lock()
+		addr := sv.state.GetPeerAddr(msg.To)
+		sv.stateMu.Unlock()
+		if addr == "" {
+			log.Printkv(context.Background(), "no-addr-for-peer", msg.To)
+			continue
+		}
+		sendmsg(addr, data)
+	}
+}
+
+// best effort. if it fails, oh well -- that's why we're using raft.
+func sendmsg(addr string, data []byte) {
+	resp, err := http.Post("http://"+addr+"/raft/msg", contentType, bytes.NewReader(data))
```

Comment:
This is gonna need to support TLS. See d49c4c2d7b0133d9f94e337c5bcb298ce89070e5. This will be simpler once @boymanjor's TLS-everywhere change goes in, then it can be hard-coded to https.